### PR TITLE
Export resize and support batch size

### PR DESCRIPTION
**Pull request by stu1130**
## Description ##
1. move the logic from python to C++ and change the Resize as HybridBlock
2. support batch_size for resize
3. add unit tests
4. support resize GPU with Hang implementation

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
@sandeep-krishnamurthy 
@zhanghang1989 
@zhreshold

### CODE REVIEW

File: python/mxnet/gluon/data/vision/transforms.py [hunk 1]
```diff
@@ -279,41 +279,29 @@ class Resize(Block):
 
 
     Inputs:
-        - **data**: input tensor with (Hi x Wi x C) shape.
+        - **data**: input tensor with (Hi x Wi x C) or (N x H x W x C) shape.
```
Comment (sandeep-krishnamurthy) [inline python/mxnet/gluon/data/vision/transforms.py]: H x W x C to be consistent?

File: src/operator/contrib/bilinear_resize.cu [hunk 2]
```diff
@@ -181,9 +108,10 @@ void SpatialUpSamplingBilinearUpdateOutput(mshadow::Stream<gpu> *s,
   dim3 blocks(static_cast<int>(num_kernels / num_threads) + 1);
   dim3 threads(num_threads);
   cudaStream_t stream = mshadow::Stream<gpu>::GetStream(s);
+  ImageLayout layout = NCHW;
```
Comment (sandeep-krishnamurthy) [inline src/operator/contrib/bilinear_resize.cu]: so layout will always be NCHW for Resize?
Shouldn't it be NHWC? Because users do imread(..) and the result NDArray will be in HWC format.

File: src/operator/image/image_utils.h [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    dst_h = src_h;
+  }
+  if (src_w < dst_w) {
+    dst_h = static_cast<int>((dst_h * src_w) / dst_w);
+    dst_w = src_w;
+  }
+  return SizeParam(dst_h, dst_w);
+}
+
+inline void CropImpl(const std::vector<TBlob> &inputs,
```
Comment (sandeep-krishnamurthy) [inline src/operator/image/image_utils.h]: Can we separate Crop implementation in a different PR? This will help to keep this PR targeted at one operation - Resize

File: src/operator/image/resize-inl.h [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                   const OpContext &ctx,
+                   const std::vector<TBlob> &inputs,
+                   const std::vector<OpReqType> &req,
+                   const std::vector<TBlob> &outputs) {
+  CHECK_EQ(outputs.size(), 1U);
+  const ResizeParam& param = nnvm::get<ResizeParam>(attrs.parsed);
+  SizeParam size;
+#if MXNET_USE_CUDA
+  if (std::is_same<xpu, gpu>::value) {
+    CHECK(param.interp == 1) << "GPU version only support bilinear interpolation";
```
Comment (sandeep-krishnamurthy) [inline src/operator/image/resize-inl.h]: Will be useful to also say, interp should be 1 for using Resize on GPU.

File: src/operator/contrib/bilinear_resize.cu (hunk 2 above)
Comment (stu1130) [reply src/operator/contrib/bilinear_resize.cu]: for bilinear resize operator, it's part of network instead of transformation. So it's always NCHW.
